doc/pinfo.1
.TH PINFO 1 "01 Dec 2001"
.SH NAME
.B pinfo
\- curses based lynx-style info browser
.SH SYNTAX
.B pinfo 
[\fIoptions\fR]
[\fBinfopage\fR]
.SH DESCRIPTION
This is a program for viewing info files. You specify which page you want to
read by passing it an
.I infopage
argument. This argument contains the name of an info page (i.e. 'bash').
The program will then (by default) search for it in 
.IR "./", 
.IR "/usr/info",
.IR "/usr/share/info",
and
.IR "/usr/local/info".
The searchpath can be adjusted by INFOPATH environment variable or in config
file. Pinfo will also automaticaly add the suffix '-info', '-info.Z', '-info.gz', or '-info.bz2'.
At present other suffixes are not recognized, but you can easily add them to
the function 
.I openinfo() 
in 
.IR "filehandling_functions.c".
.P
When the search for info pages fails, man is called with the infopage
argument, and it's output is parsed by pinfo. This means that when you don't
have the appropriate info page, but have a man page instead; the man page
will be viewed.
.P
When no \fIinfopage\fR is specified, the default `dir' page is shown.
.P
Supported options are
.P
.BR "-h", 
.B --help
\- print help information and exit.
.P
.BR "-v", 
.BR --version
\- print version information and exit.
.P
.BR "-m",
.BR --manual 
\- uses manual page instead of info by default. (pinfo -m could be used as a
manual pager). \fBWarning\fR: Everything what follows this option is passed
to the `\fIman\fR' program. Don't be confused if pinfo options, which
followed `\fB-m\fR' don't work. When using this option, pinfo does not parse
the info options as ussual! It invokes the man part of program.
.P
You can also call the man function of pinfo in another way. When pinfo is
called with an argv[0] (the program file name), which contains the word 'man'
in it's name, the man functions are enabled automatically.
.P
Previously there was a symlink to pinfo, called pman, but I had to remove it
from the distribution, since it's name was in conflict with some other
utility. Anyway, you can feel free to create such a link if you wish.
.P
.BR "-r",
.BR --raw-filename 
\- uses a raw filename first (i.e. the name which you specified as
\fIinfopage\fR is considered to be a real file in the specified location).
.P
.BR "-f",
.BR --file 
synonym for -r.
.P
.BR "-a",
.BR --apropos 
\- if this is set, apropos is called when no man or info page could be found.
.P
.BR "-c",
.BR --cut-man-headers 
\- if this is set, man parsing code will try to cut out the repeated man
headers. Use with care. ;)
.P
.BR "-s",
.BR "--squeeze-lines" \-
cut empty lines from manual pages. This option enables autocutting of every
repeated newline in a manual page.
.P
.BR "-t",
.BR "--force-manual-tag-table" \-
forces manual detection of tag table. This allows you to view info pages,
which may be corrupted. (as i.e. version of jed's pages, shipped with
RH5.0). The tag table corruption ussualy appears in that the info links,
which you follow, move you to quite unexpected nodes.
.P
.BR "--node=\fInodename\fB",
.BR "--node \fInodename\fB" \-
Go to the node `\fInodename\fR' of info file. Since 0.6.7 it is also
possible to specify nodes as in standalone info via filenames, like
`(gcc)Introduction'.
.P
.BR "--rcfile=\fIfilename\fB",
.BR "--rcfile \fIfilename\fB" \-
Use alternate rcfile.
.P
.BR "--long-manual-links",
.BR "-l" \-
Use long link names in manuals. On some systems the manual hierarchy is
divided into subsections like `3ncurses', etc, while on other systems all
belongs to section `3'. If this option is what your system is like, feel
free to use it.
.P
.BR "--clear-at-exit",
.BR "-x" \-
Clear screen at exit.
.P
The options are handled by GNU getopt, so you can here (as in other
programs) abbreviate the option names to the minimal number of
characters by which the options differ.
.P
Warning! If you do not have getopt, these options will not work!
.P
.SH DEFAULT KEYS WHEN BROWSING INFO FILE
.P
Just take a look at the example config file (below), and at the key 
descriptions. Keys available in manual viewer differ a bit from the keys 
available in info viewer.
.P
.SH ENVIRONMENT
There is a variable $INFOPATH, which can specify the paths to be searched
for info files. It's format is similar to that of the $PATH variable. An
example setting could look like:
.P
.B /usr/info:/usr/somewhere/info:/not/even/in/usr/info
.P
etc. Directories are separated by colons.

.SH COLOR AND KEY DEFINITIONS
.P
There are configuration files called ~/.pinforc and
[prefix]/etc/pinforc, for local and global configuration (where prefix is the
prefix of the directory, where pinfo is installed, i.e. /usr/local, or /).
Here's an example of such a file; we'll discuss the contents below:
.P
.nf
.sp
# Here are some colour setting.
# Whitespace between the entries is optional.
COL_NORMAL = COLOR_WHITE, COLOR_BLACK, NO_BOLD, NO_BLINK
COL_MENUSELECTED = COLOR_RED, COLOR_BLACK, BOLD, NO_BLINK
COL_MENU=COLOR_BLUE,COLOR_BLACK,BOLD, NO_BLINK
COL_NOTESELECTED=COLOR_RED,COLOR_BLACK,BOLD, NO_BLINK
COL_NOTE=COLOR_GREEN,COLOR_BLACK,BOLD, NO_BLINK
COL_TOPLINE=COLOR_YELLOW,COLOR_BLUE,BOLD, NO_BLINK
COL_BOTTOMLINE=COLOR_YELLOW,COLOR_BLUE,BOLD, NO_BLINK
COL_MANUALBOLD=COLOR_WHITE,COLOR_BLACK,BOLD, NO_BLINK
COL_MANUALITALIC=COLOR_WHITE,COLOR_BLACK,BOLD, NO_BLINK
COL_URL=COLOR_MAGENTA,COLOR_BLACK,BOLD, NO_BLINK
COL_URLSELECTED=COLOR_RED,COLOR_BLACK,NO_BOLD, NO_BLINK
COL_INFOHIGHLIGHT=COLOR_WHITE,COLOR_BLACK,BOLD, NO_BLINK
#
# Here are some keybindings as well...
#
KEY_TOTALSEARCH_1 = 's'
KEY_TOTALSEARCH_2 = 'S'
KEY_SEARCH_1 = '/'
KEY_SEARCH_2 = '.'
KEY_GOTO_1='g'
KEY_GOTO_2='m'
KEY_HOME_1='h'
KEY_HOME_2='H'
KEY_PREVNODE_1='p'
KEY_PREVNODE_2='P'
KEY_NEXTNODE_1='n'
KEY_NEXTNODE_2='N'
KEY_UP_1=KEY_UP
KEY_UP_2='u'
KEY_END_1=KEY_END
KEY_END_2='e'
KEY_PGDN_1=KEY_NPAGE
KEY_PGDN_2=' '
KEY_PGDN_AUTO_1=0
KEY_PGDN_AUTO_2=' '
KEY_PGUP_1=KEY_PPAGE
KEY_PGUP_2='b'
KEY_PGUP_AUTO_1=0
KEY_PGUP_AUTO_2='b'
KEY_DOWN_1=KEY_DOWN
KEY_DOWN_2='d'
KEY_TOP_1=KEY_HOME
KEY_TOP_2='t'
KEY_BACK_1=KEY_LEFT
KEY_BACK_2='l'
KEY_FOLLOWLINK_1=KEY_RIGHT
KEY_FOLLOWLINK_2='\n'
# 12 is a code for ctrl+l
KEY_REFRESH_1=12
KEY_REFRESH_2='~'
KEY_SHELLFEED_1='!'
KEY_SHELLFEED_2='1'
KEY_QUIT_1='q'
KEY_QUIT_2='Q'
KEY_DIRPAGE_1='d'
KEY_DIRPAGE_2='D'
KEY_GOLINE_1='l'
KEY_GOLINE_2=0
KEY_PRINT_1=']'
KEY_PRINT_2=0
#
# Some options, explained in the man page
#
MANUAL=false
CUT-MAN-HEADERS=true
CUT-EMPTY-MAN-LINES=true
RAW-FILENAME=false
APROPOS=false
DONT-HANDLE-WITHOUT-TAG-TABLE=false
LONG-MANUAL-LINKS=false
FILTER-0xB7=true
QUIT-CONFIRMATION=false
QUIT-CONFIRM-DEFAULT=no
CLEAR-SCREEN-AT-EXIT=true
STDERR-REDIRECTION="2> /dev/null"
HTTPVIEWER=lynx
FTPVIEWER=lynx
MAILEDITOR=pine
MANLINKS=1:8:2:3:4:5:6:7:9:n:p:o:3X11:3Xt
INFOPATH=/usr/info:/usr/share/info:/usr/local/info
HIGHLIGHTREGEXP=Bash.*has
SAFE-USER=nobody
SAFE-GROUP=nobody
.fi
.P
As you can see, the format is simple. First I'll explain the color
definitions. First you must enter a color name (all available color
names are present in the example, and they're self explanatory, I
think. There is also a special color COLOR_DEFAULT, which stands for
transparency). Then you enter the foreground color, and the background
color. The BOLD attribute means that we want the foreground color to
be highlighted. (i.e.  lightblue, lightgreen). BLINK attribute is the
blinking attribute, or highlighted background in some other configurations.
.P
Now let's move to the key definitions. Here we first put a key name (again
all keys are present in the example); then we enter it's value -- either
surrounded
by apostrophes, or a keycode number (like in KEY_REFRESH_1), or its mnemonic
code name if it's a special key (like i.e. in KEY_FOLLOWLINK_1).
.P
If you wish to specify key by code value, use the supplied program 'testkey'
to obtain the needed value. It mainly is a feature, when you want to add
some ctrl+letter keybindings, and similar.
.P
For each function you can bind two keys, i.e. you could bind both
Enter and Cursor Right to the FollowLink-function.  As you can see in
the example above, the two key names are KEY_FOLLOWLINK_1 and
KEY_FOLLOWLINK_2.  
.P
Here's an explanation of the key names:
.RS
.sp
.TP 1.0i
.B KEY_TOTALSEARCH_1
Key for searching through all nodes of info file.
.TP 
.B KEY_TOTALSEARCH_2
Alternate key for searching through all nodes of info file.
.TP 
.B KEY_SEARCH_1
Key for searching through current node (or manual).
.TP 
.B KEY_SEARCH_2
Alternate key for searching through current node (or manual).
.TP 
.B KEY_SEARCH_AGAIN_1
Key for repeating the last search.
.TP 
.B KEY_SEARCH_AGAIN_2
Alternate key for repeating the last search.
.TP
.B KEY_GOTO_1
Key for explicitly going to a node (by specifing it's name).
.TP 
.B KEY_GOTO_2
Alternate key for explicitly going to a node (by specifing it's name).
.TP 
.B KEY_PREVNODE_1
Key for going to a node marked as 'Prev' in the header. In manpage viewer
this goes to the previous man section.
.TP 
.B KEY_PREVNODE_2
Alternate key for going to a node marked as 'Prev' in the header. In manpage
viewer this goest to the previous man section.
.TP 
.B KEY_NEXTNODE_1
Key for going to a node marked as 'Next' in the header. In manpage viewer
this goes to the next man section.
.TP 
.B KEY_NEXTNODE_2
Alternate key for going to a node marked as 'Next' in the header. In manpage
viewer this goes to the next man section.
.TP 
.B KEY_UP_1
Key for scrolling text one line up.
.TP 
.B KEY_UP_2
Alternate key for scrolling text one line up.
.TP 
.B KEY_END_1
Key for going to the end of the node.
.TP 
.B KEY_END_2
Alternate key for going to the end of the node.
.TP 
.B KEY_PGDN_1
Key for going one page down in the viewed node.
.TP 
.B KEY_PGDN_2
Alternate key for going one page down in the viewed node.
.TP
.B KEY_PGDN_AUTO_1
Key for going to the next node when you're at the end of node (default
is zero -- turned off).
.TP
.B KEY_PGDN_AUTO_2
Alternate key for going to the next node when you're at the end of node
(default is space, as for pgdn_2).
.TP 
.B KEY_HOME_1
Key for going to the beginning of the node.
.TP 
.B KEY_HOME_2
Alternate key for going to the beginning of the node.
.TP 
.B KEY_PGUP_1
Key for going one page up in the viewed node.
.TP 
.B KEY_PGUP_2
Alternate key for going one page up in the viewed node.
.TP
.B KEY_PGUP_AUTO_1
Key for going to the `up' node, when being at the top of node. (Default value
is zero -- turned off).
.TP
.B KEY_PGUP_AUTO_2
Alternate key for going to the `up' node, when being at the top of node.
(Default value is `\-', as for pgup_2).
.TP 
.B KEY_DOWN_1
Key for scrolling the text down one line.
.TP 
.B KEY_DOWN_2
Alternate key for scrolling the text down one line.
.TP 
.B KEY_TOP_1
Key for going to the top (first) node.
.TP 
.B KEY_TOP_2
Alternate key for going to the top (first) node.
.TP 
.B KEY_BACK_1
Key for going back (in the history of viewed nodes).
.TP 
.B KEY_BACK_2
Alternate key for going back (in the history of viewed nodes).
.TP 
.B KEY_FOLLOWLINK_1
Key for following a hypertext link.
.TP 
.B KEY_FOLLOWLINK_2
Alternate key for following a hypertext link.
.TP 
.B KEY_REFRESH_1
Key for refreshing the screen (hardcoded is the ^L value).
.TP 
.B KEY_REFRESH_2
Alternate key for refreshing the screen.
.TP 
.B KEY_SHELLFEED_1
Key for calling a shell command, and passing the viewed node to the stdin of

NEWS
0.6.8
	Patches from Witold Filipczyk for node detection algorithms in man
	and info.
	Added default item on search prompt.
	Bugfix in info link freeing procedure, which failed on exim pages.
	Bugfix for dir page lookup when several matches are found--there was
	a double close() problem.
0.6.7	
	Added scrolling in left/right direction (!!!new config file keys).
	Fixed handling of info entries like `emacs21/emacs'--modified
        function matchfile().
	Added autmatical detection of rawfilename for filenames beginning
	with `../', `./', `/' in main().
	Added support for '(gcc)Introduction' like specification of nodes.
0.6.6	
	bugfixes for userdefined rcfile and configuration inconsistences
	added sorting of matches for the info file to display
0.6.5	
	Fixed bug in sorting of tag table links (major effect to the engine)
	Changed the file search engine to scan the directory for a matching
	pattern (instead of trying various propositions).
	Fixed minor sigsegv bug in page loader (link initializer).
	!Added switching between sections in manualviewer mode
	Highlighting code does not highlight leading spaces in splitted
	notes anymore.
	Added information when the search is not successful
	Added support for command line selected rc files
0.6.4	
	New apropos handler. It works now interactively. New option -p here.
	Doesn't highlight anymore links to itself in man viewer.
	Fixed bug with link following.
        Fixed readline emulation: it now deletes old line content,
        and handles end key
        Fixed SIGWINCH handler/curses blocking getch emulation
	Fixed searchpath routines. First of all, there is introduced a
	'context' for currently viewed file. Pinfo will search files for
	one set of infopages only in one directory--in the context directory.
	The context can be changed only when changing the set of infopages.
0.6.3   
	NLS and documentation bugfixes
	Small bugfix in keyhandler
0.6.2	
	Fixed bug with searchpath for raw filename specification (temporary
	hack).
	Fixed stupid bug with info parser, which didn't accept an
	`INFOTAG.+\n' separator for info nodes (eg. gnuplot's infopage).
	Fixd detection of first info page.
	Fixed buggy use of bkgdset()
0.6.1   
	Modified behaviour of auto-pgdn. It now expands the submenus, as in
	info.
	Fixed problem with pinforc installed to non-exsisting etc directory
	Fixed problem with black bar on first man page
	Fixed buggy whitespace-skip comparing procedure 
	Manual code can now handle hyphenated url links (it's still not
	perfect with e-mail--it only works if the `@' is in the upper line)
	Fixed buf in indirect-file search engine (a misplaced if())
	Added texinfo documentation.
	Manual code is now case insensitive as far as sections are considered
0.6.0	
	Replaced evil `myclrtoeol' with `bkgdset+clrtoeol'
	Fixed putenv handler
	Make install does not overwrite old pinforc anymore (this has it's
	bad side--you must deal alone with version incompatibilities)
	fixed sorting of dirpage matches
	fixed some small bugs
0.5.9
	Setenv()->putenv() (so pinfo can now compile without problems
	at nonlinux unices)
	Added SAFE-GROUP, and SAFE-USER config file options for defining
	users, to which we want to drop root privileges when running pinfo
	Added HIGHLIGHTREGEXP option, to allow highlighting of user defined
	regexps when viewing manuals/infos
	Search highlighting works now "professionaly" (as in less)
	Fixed sigsegv bug in search engine.
0.5.8
	Fixed some ugly bugs to prepare for world domination ;->
0.5.7	Updated man page
	Fixed `splitted manlink' code
	Added key_goto to manual code
	Fixed minor bugs, and portability misadvantages
	Modified the default info search order (./ is the last dir)
	Manual code displays now manuals with optional width != than 80 chars.
	Fixed configure --with-included-gettext (hopefuly)
	-r works now with compressed files
0.5.6	
	Fixed memory bugs
	Fixed displaying of concatenated dir pages
	Speed up alghoritms
	Transparent colors
	Manual code can now handle hyperlinks splitted by `-' sign
	Security filter for dangerous man macros (experimental)
0.5.5	
	Fixed nasty sigsegv bug in manual code.
	Speed improvements to the code.
	Many memory bugs fixed in the whole code.
	Added key for going to specified line.
	Added dir page name lookups. Now you will be able f.i. to go to 
	`textutils:cat'	after typing `pinfo cat'.
	Added support for printing. (with highlights, as in groff -Tascii).
0.5.4
	Fixed highlighting code (the handling of \t's).
	PgDn works now more `traditionaly' -- at end the last link is
	selected.
	Added FILTER-0xB7 config option.
	Fixed bug in `*note:' detection.
	Added option for enabling quit confirmation at exit (as in lynx).
	Added option for clearing screen at exit
	Added support for lightbackground
	Fixed nonblack background in displaying contents of text.
	Added support for multi-dired info dir entries.
	Finally done some simple search highlighting.
0.5.3
	Fixed stderr redirections. Now you won't get garbage on screen when
	following manual links.
	Added MAN-OPTIONS config option, which allows you to pass options
	to te `man' program.
	Added KEY_CTRL and KEY_ALT macros, which enable you to use CTRL+x
	and ALT+x combinations. (the ALT+x is probably only usable on linux
	console, but who knows ;)
	Fixed global regexp search code. It now searches every two following
	lines, to avoid returning of expressions, which are several hounderts
	lines long ;)
	Added STDERR-REDIRECTION config file option, which allows you to
	specify how do you wish to redirect stderr from plug-ins.

src/pinfo.c
/******************************************************************************
* Pinfo is a lynx-style info and manual viewer. It is written by              *
* Przemek Borys <pborys@dione.ids.pl>. Copying policy is GPL                  *
******************************************************************************/

#include "common_includes.h"

RCSID (PKG_VER "$Id$")

#ifdef HAVE_GETOPT_LONG
#include <getopt.h>
#endif

     char *version = VERSION;
     int DontHandleWithoutTagTable = 0;

     char *curfile = 0;		/* currently viewed filename */

     char *pinfo_start_node = 0;	/* node specified by --node option */

     void strip_file_from_info_suffix (char *file);	/* 
							 * strip `.info' suffix from
							 * "file" 
							 */
     char *addinfosuffix (char *file);	/* add `.info' suffix to "file" */

     void checksu ();		/* protect against bad, bad macros */

     int
       main (int argc, char *argv[])
{
  int filenotfound = 0;
  char filename[256];
  WorkRVal work_return_value =
  {0, 0};
  int i, userdefinedrc = 0;
  int command_line_option;
  FILE *id = NULL;
  long lines = 0;		/* line count in message */
  char **message = 0;		/* this will hold node's text */
  char *type = 0;		/* this will hold the node's header */
  int tag_table_pos = 1;
  char *tmp;
#ifdef HAVE_GETOPT_LONG
  static struct option long_options[] =
  {
    {"help", 0, 0, 'h'},
    {"version", 0, 0, 'v'},
    {"manual", 0, 0, 'm'},
    {"file", 0, 0, 'f'},
    {"raw-filename", 0, 0, 'r'},
    {"apropos", 0, 0, 'a'},
    {"plain-apropos", 0, 0, 'p'},
    {"cut-man-headers", 0, 0, 'c'},
    {"squeeze-manlines", 0, 0, 's'},
    {"dont-handle-without-tag-table", 0, 0, 'd'},
    {"force-manual-tag-table", 0, 0, 't'},
    {"node", 1, 0, 'n'},
    {"long-manual-links", 0, 0, 'l'},
    {"clear-at-exit", 0, 0, 'x'},
    {"rcfile", 1, 0, 1},	/* no one-letter shortcut :( */
    {0, 0, 0, 0}};
#endif
  signal_handler ();		/* take care of SIGSEGV, SIGTERM, SIGINT */
  searchagain.type = 0;
  searchagain.search = 0;
  initlocale ();
  inithistory ();
  for (i = 1; i < argc; i++)
    if (strncmp (argv[i], "--rcfile", 8) == 0)
      userdefinedrc = 1;
  if (!userdefinedrc)
    parse_config ();		/* read config information */
  if (verbose)
    printf ("Przemek's Info Viewer v%s\n", version);
  if (argc == 1)		/* if no arguments were given */
    {
      id = openinfo ("dir", 0);
      curfile = xmalloc (150);
      strcpy (curfile, "dir");
      strcpy (filename, "dir");
    }
  if ((strlen (argv[0]) >= 3)||(use_manual))
    if ((strstr (argv[0], "man") != NULL)||(use_manual))	/* handle any 'man' alias to 'pinfo' */
      {
	if (verbose)
	  printf (_ ("Looking for man page...\n"));
	strcpy (filename, "");
	for (i = 1; i < argc; i++)	/* pass all arguments to the `man' 
					   command (manhandler calls `man') */
	  {
	    strcat (filename, argv[i]);
	    strcat (filename, " ");
	  }
	exit (handlemanual (filename));
      }



#ifdef HAVE_GETOPT_LONG

/******************************************************************************
* Parse command line options (getopt)                                         *
******************************************************************************/

  do
    {
      command_line_option = getopt_long (argc, argv, "hvmfrapcsdtnlx", long_options, NULL);
      switch (command_line_option)
	{
	case 'x':
	  ClearScreenAtExit = 1;
	  break;
	case 'l':
	  LongManualLinks = 1;
	  break;
	case 'n':
	  if (!optarg)
	    {
	      printf (_ ("--node option used without argument\n"));
	      exit (1);
	    }
	  pinfo_start_node = malloc (strlen (optarg) + 1);
	  strcpy (pinfo_start_node, optarg);
	  break;
	case 1:		/* rcfile */
	  if (!optarg)
	    {
	      printf (_ ("--rcfile option used without argument\n"));
	      exit (1);
	    }
	  rcfile = strdup (optarg);
	  parse_config ();	/* parse user-defined config file */
	  break;
	case 't':
	  ForceManualTagTable = 1;
	  break;
	case 'h':
	  printf (_ ("Usage:\n" \
		     "%s [options] [info|manual]\n" \
		     "Options:\n" \
		     "-h, --help                            help\n" \
		     "-v, --version                         version\n" \
		     "-m, --manual                          use man page\n" \
		"-r, --raw-filename                    use raw filename\n" \
		  "-f, --file                            synonym for -r\n" \
		     "-a, --apropos                         call apropos if nothing found\n" \
	       "-p, --plain-apropos                   call only apropos\n" \
		     "-c, --cut-man-headers                 cut out repeated man headers\n" \
		     "-l, --long-manual-links               use long link names in manuals\n" \
		     "-s, --squeeze-manlines                cut empty lines from manual pages\n" \
		     "-d, --dont-handle-without-tag-table   don't display texinfo pages without tag\n" \
		     "                                      tables\n" \
		     "-t, --force-manual-tag-table          force manual detection of tag table\n" \
	    "-x, --clear-at-exit                   clear screen at exit\n" \
		     "    --node=nodename, --node nodename  jump directly to the node nodename\n" \
	    "    --rcfile=file, --rcfile file      use alternate rcfile\n"),
		  argv[0]);
	  exit (0);
	case 'v':
	  exit (0);
	case 'm':
	  checksu ();
	  if (verbose)
	    printf (_ ("Looking for man page...\n"));
	  strcpy (filename, "");
	  for (i = optind; i < argc; i++)
	    {
	      strcat (filename, argv[i]);
	      strcat (filename, " ");
	    }
	  exit (handlemanual (filename));
	case 'f':
	case 'r':
	  strncpy (filename, argv[argc - 1], 200);
	  checkfilename (filename);	/* security check */
	  addrawpath (filename);	/* add the raw path to searchpath */
	  tmp = filename + strlen (filename) - 1;
	  strip_compression_suffix (filename);	/* later, openinfo automaticaly adds them */
	  while ((tmp > filename) && (*tmp != '/'))
	    tmp--;		/* get basename */
	  if (*tmp == '/')
	    tmp++;
	  id = openinfo (tmp, 0);	/* and try it without '.info' suffix */
	  break;
	case 'a':
	  use_apropos = 1;
	  break;
	case 'p':
	  use_apropos = 1;
	  plain_apropos = 1;
	  strncpy (filename, argv[argc - 1], 200);
	  exit (handlemanual (filename));
	  break;
	case 'c':
	  CutManHeaders = 1;
	  break;
	case 'd':
	  DontHandleWithoutTagTable = 1;
	  break;
	case 's':
	  CutEmptyManLines = 1;
	  break;
	}
    }
  while (command_line_option != EOF);
/*****************************************************************************/
#endif

  checksu ();
  initpaths ();

  if (argc > 1)
    {
#ifdef HAVE_GETOPT_LONG
      if (optind < argc)
	{
	  strncpy (filename, argv[optind], 200);	/* the paths will be searched
							   by openinfo() */
	}
      else
	{
	  strcpy (filename, "dir");
	}

#else
      strncpy (filename, argv[argc - 1], 200);	/* the paths will be searched
						   by openinfo() */
#endif
      if(filename[0]=='(')
        {
          int fnamelen=strlen(filename);
          for(i=0;i<fnamelen;i++)	/* erase the leading '(' */
            filename[i]=filename[i+1];
          for(i=0;filename[i]!=')';i++);
          filename[i]=0;		/* leave the filename part in filename */
          if(!pinfo_start_node)		/* copy the node content to pinfo_start_node */
            {
              pinfo_start_node=strdup(&filename[i+1]);
            }
        }

      checkfilename (filename);	/* security check */

      if((strncmp(filename,"../",3)==0)||	/* autodetect raw filenames */
         (strncmp(filename,"./",2)==0)||
         (filename[0]=='/'))
         {
           addrawpath(filename);
         }
               
      curfile = xmalloc (strlen (filename) + 100);	/* leave some space for
							   `.info' suffix */
      strcpy (curfile, filename);
    }

  if(id == NULL)		/* no rawpath has been opened */
    id = openinfo (filename, 0);

  if (id == NULL)		/* try to lookup the name in dir file */
    {
      id = dirpage_lookup (&type, &message, &lines, filename, &pinfo_start_node);
    }
  if (id == NULL)		/* if still nothing, try to use man page instead */
    {
      printf (_ ("Error: could not open info file, trying manual\n"));
      exit (handlemanual (filename));
    }
  if (seek_indirect (id))	/* search for indirect entries, if any */
    {
      read_item (id, &type, &message, &lines);
      load_indirect (message, lines);
    }


  if (seek_tag_table (id,1) != 2)	/* load tag table if such exists... */
    {
      if (ForceManualTagTable == 0)
	{
	  read_item (id, &type, &message, &lines);
	  load_tag_table (message, lines);
	}
      else
	{
	  if (indirect)
	    create_indirect_tag_table ();
	  else
	    {
	      fseek (id, SEEK_SET, 0);
	      create_tag_table (id);
	    }
	}
    }
  else
    /* ...otherwise try to create one */
    {
      if ((verbose)&&(strcmp(curfile,"dir")))
	printf (_ ("Warning: tag table not found...\n"));
      if (!DontHandleWithoutTagTable)
	{
	  if ((verbose)&&(strcmp(curfile,"dir")))
	    printf (_ ("Trying to create alternate tag table...\n"));
	  create_tag_table (id);
	  if (TagTableEntries < 1)	/* if there weren't found any info
					   entries */
	    {
	      printf (_ ("This doesn't look like info file...\n"));
	      exit (handlemanual (filename));
	    }
	}
      else
	return 1;
    }

  if (pinfo_start_node)
    {
      tag_table_pos = gettagtablepos (pinfo_start_node);
      if (tag_table_pos == -1)
	{
	  printf (_ ("Specified node does not exist...\n"));
	  return 1;
	}
    }
  else
    {
      tag_table_pos = gettagtablepos (FirstNodeName);
    }
  init_curses ();		/* initialize curses screen interface */

  do
    {
      seeknode (tag_table_pos, &id);	/* set seek offset for given node */
      read_item (id, &type, &message, &lines);	/* read the node */

      if (!filenotfound)	/* handle goto/link where no file was found 
				   -- see bellow */
	addinfohistory (curfile, tag_table[tag_table_pos].nodename, -1, -1, -1);
      else
	filenotfound = 0;
      work_return_value = work (&message, &type, &lines, id, tag_table_pos);
      if (work_return_value.node)
	{
	  if (work_return_value.file[0] == 0)	/* no cross-file link selected */
	    {
	      int tmppos = gettagtablepos (work_return_value.node);
	      if (tmppos != -1)
		tag_table_pos = tmppos;
	    }
	  else
	    /* file was specified */
	    {
	      strip_file_from_info_suffix (work_return_value.file);
	      if (strcmp (curfile, work_return_value.file) == 0)
		/* file name was the same with the file currently viewed */
		{
		  int tmppos = gettagtablepos (work_return_value.node);
		  if (tmppos != -1)
		    tag_table_pos = tmppos;
		}
	      else
		/* open new info file */
		{
		  char *tmp;
		  fclose (id);
		  tmp = strdup (work_return_value.file);	/*tmp = addinfosuffix (work_return_value.file); */
		  clearfilenameprefix ();
		  id = openinfo (tmp, 0);
		  xfree (tmp);
		  tmp = 0;
		  if (id == NULL)	/* if the file doesn't exist */
		    {
		      attrset (bottomline);
		      mvhline (maxy - 1, 0, ' ', maxx);
		      mvaddstr (maxy - 1, 0, _ ("File not found. Press any key..."));
		      move (0, 0);
		      attrset (normal);
		      getch ();
		      filenotfound = 1;
		      if (infohistory.length)
			{
			  npos = infohistory.pos[infohistory.length];
			  ncursor = infohistory.cursor[infohistory.length];
			}
		      strip_file_from_info_suffix (curfile);	/* open back the old file */
		      tmp = strdup (curfile);	/*tmp = addinfosuffix (curfile); */
		      id = openinfo (tmp, 0);
		      xfree (tmp);
		      tmp = 0;
		      if (id == NULL)
			{
			  closeprogram ();
			  printf (_ ("Unexpected error.\n"));
			  return 1;
			}
		    }
		  else
		    /* if we succeeded in opening new file */
		    {
		      if (curfile)
			{
			  xfree (curfile);
			  curfile = 0;
			}
		      curfile = xmalloc (strlen (work_return_value.file) + 150);
		      strcpy (curfile, work_return_value.file);
		      freeindirect ();
		      if (seek_indirect (id))	/* find the indirect entry */
			{
			  read_item (id, &type, &message, &lines);	/* read it */
			  load_indirect (message, lines);	/* initialize indirect entries */
			}
		      freetagtable ();	/* free old tag table */
		      if (seek_tag_table (id,0) != 2)	/* search for the new tagtable */
			{
			  if (ForceManualTagTable == 0)		/* if no manual initialization requested, load the tag table */
			    {
			      read_item (id, &type, &message, &lines);
			      load_tag_table (message, lines);
			    }
			  else
			    /* create tag table manually */
			    {
			      if (indirect)
				create_indirect_tag_table ();
			      else
				{
				  fseek (id, SEEK_SET, 0);
				  create_tag_table (id);
				}
			    }
			}
		      else
			/* no tagtable found */
			{
			  if (!DontHandleWithoutTagTable)
			    {
			      TagTableEntries = 0;
			      mvhline (maxy - 1, 0, ' ', maxx);
			      mvaddstr (maxy - 1, 0, _ ("Tag table not found. Trying to create alternate..."));
			      create_tag_table (id);
			      if (TagTableEntries < 1)
				{
				  closeprogram ();
				  printf (_ ("This doesn't look like info file...\n"));
				  return 1;
				}
			    }
			  else
			    return 1;
			}
		      if (work_return_value.node[0] != 0)
			{
			  int tmptagtablepos = gettagtablepos (work_return_value.node);
			  if (tmptagtablepos != -1)
			    tag_table_pos = tmptagtablepos;
			  else
			    tag_table_pos = gettagtablepos (FirstNodeName);
			}
		      else
			tag_table_pos = gettagtablepos (FirstNodeName);

		    }		/* end: open new info file -- file exists */
		}		/* end: open new info file */
	    }			/* end: file name was specified */
	}			/* end: node was specified in work return value */
    }
  while (work_return_value.node);
  fclose (id);
  closeprogram ();
  freelinks ();			/* free's at the end are optional, but look nice :) */
  freeitem (&type, &message, &lines);
  freetagtable ();
  freeindirect ();
  return 0;
}
void
strip_file_from_info_suffix (char *file)
{
  if (strlen (file) > 5)
    {
      if (strcmp (file + strlen (file) - 5, ".info") == 0)
	{
	  file = file + strlen (file) - 5;
	  *file = 0;
	}
    }
}
char *
addinfosuffix (char *info)
{
  char *withsuffix = xmalloc (strlen (info) + 150);
  strcpy (withsuffix, info);
  if (strlen (info) == 3)
    {
      if (strcmp ("dir", info) != 0)
	strcat (withsuffix, ".info");
    }
  else
    strcat (withsuffix, ".info");

  return withsuffix;
}

/* If pinfo was called by root then it should work as nobody.   */
/* This protect us against .pso and .open macros which could    */
/* be used for breaking the system's security.                  */

void
checksu ()
{
  struct passwd *pswd;
  struct group *grwd;

  if (!getegid () || !getgid ())
    {
      grwd = getgrnam (safe_group);
      if (!grwd)
	{
	  if (verbose)
	    {
	      printf (_ ("Security warning: Unable to get GID of group called: %s\n"), safe_group);
	      sleep (1);
	    }
	}
      else
	{
	  if (!getgid () && !getuid ())
	    setgid (grwd->gr_gid);
	  else
	    setegid (grwd->gr_gid);
	}
    }

  if (!geteuid () || !getuid ())
    {
      pswd = getpwnam (safe_user);
      if (!pswd)
	{
	  if (verbose)
	    {
	      printf (_ ("Security warning: Unable to get UID of user called: %s\n"), safe_user);
	      sleep (1);
	    }
	}
      else
	{
	  if (!getuid ())
	    setuid (pswd->pw_uid);
	  else
	    seteuid (pswd->pw_uid);
	}
    }

}
